### ISSUE TITLE
[Bug Report] Incorrect rounding when using `bfp8` in `ttnn.from_torch`.

### ISSUE THREAD

**Issue by hschoi4448** [tenstorrent/tt-metal#15889]
**Describe the bug**
There is incorrect rounding when using bfp8 in `ttnn.from_torch`.

**To Reproduce**
```Python
import pytest
import torch
import ttnn


def test_example(device):
    # bfp8_b rounding issue. 
    # mantissa for 64.5 is 0000 001
    # According to the "Round to nearest, ties to even" rule, the mantissa should be 0.
    value = 64.5
    torch_input = torch.tensor([value], dtype=torch.float)
    tt_input = ttnn.from_torch(torch_input, layout=ttnn.TILE_LAYOUT, dtype=ttnn.bfloat8_b, device=device)
    tt_output = ttnn.to_torch(tt_input)

    print("python round", round(value)) # 64
    print("tt_output ", tt_output) # 65. this is not ties to even
    print("===================================")

    # bfp8_b rounding issue2, mantissa overflow was not handled properly
    # this is the case all mantissa is 1 and sholud round up
    value = 7.9999
    torch_input = torch.tensor([value], dtype=torch.float)
    tt_input = ttnn.from_torch(torch_input, layout=ttnn.TILE_LAYOUT, dtype=ttnn.bfloat8_b, device=device)
    tt_output = ttnn.to_torch(tt_input)

    torch_output = torch_input.to(dtype=torch.bfloat16)
    print("torch_output", torch_input.round()) # 8
    print("tt_output ", tt_output) # 7.9375
```

1. Run above comment
2. See result

The first case is an example where 64.5 is converted to the `bfp8_b` type, and rounding occurs. According to the "Round to nearest, ties to even" rule, 64.5 should round to 64, but for some unknown reason, it has rounded to 65.

The second case is when 7.9999 is rounded. 7.9999 should round to 8, but it seems that `bfp8` has an issue handling the case where the mantissa is all 1s, which prevents it from rounding to 8.

As a result, instead of 8, the value becomes 7.9375, leading to an error of 0.0625, when the expected error should have been 0.0001.

**Expected behavior**
The results of `ttnn.from_torch` for `bfp8` should follow the "Round to nearest, ties to even" rule.

**Please complete the following environment information:**
- OS: [e.g. Ubuntu 20.04]
- Version of software (eg. commit 23ce00c668d7bff14c98657d1245f58a65fe0cef)

**Additional context**

**Comment by ayerofieiev-tt** [comment]
Please see these anwers.
I expect this ticket can be closed.

https://github.com/tenstorrent/tt-metal/issues/15914#issuecomment-2539881112

https://github.com/tenstorrent/tt-metal/issues/14090#issuecomment-2539812874

**Comment by hschoi4448** [comment]
> Please see these anwers. I expect this ticket can be closed.
> 
> [#15914 (comment)](https://github.com/tenstorrent/tt-metal/issues/15914#issuecomment-2539881112)
> 
> [#14090 (comment)](https://github.com/tenstorrent/tt-metal/issues/14090#issuecomment-2539812874)

I’ve reviewed your response, but I believe this is a different issue.

1. I understand that `bfp8` lacks 1 bit.
The float value 64.5 requires exactly 7 bits in the mantissa, making it a number that needs rounding when converted to `bfp8`.
Currently, the `from_torch` function is not handling the rounding of 64.5 correctly.


float                            64.5, mantissa            : 0000 001  <--- 7bits. need rounding
correct bfp8 rounding 64   , mantissa in bfp8: 1000 00 <--- 6bits
actual  bfp8 rounding 65   , mantissa: in bfp8: 1000 01 <--- 6bits. not following ties to even

2. The second issue is also related to rounding.  
If the last bits were simply truncated, it could be considered part of the specification. However, the current `from_torch` implementation is attempting to perform rounding, but it is doing so incorrectly, which means the specification itself is flawed.

float             7.9999 exp and mantissa: 1000 0001, 1111 11111111111
correct bfp8 8          exp and mantissa: 1000 0010, 1000 00 <-- mantissa overflow, should increase exp
actual  bfp8  7.9375 exp and mantissa: 1000 0001, 1111 11

**Comment by zzigler-tt** [comment]
@ayerofieiev-tt Hello, checking in on the progress / next steps for this issue. TY

**Comment by TT-BrianLiu** [comment]
I ported our conversion process to C++ so you can run and play around with it:

```
#include <vector>
#include <iostream>


template <bool truncate_bfp_mantissa = false>
inline uint8_t convert_u32_to_bfp8(uint32_t input, uint32_t shared_exp, bool is_exp_a) {
    // check for both +/- 0.0
    constexpr uint32_t EXP_MANTISSA_BMSK = ((1U << 31) - 1);
    bool is_zero = ((input & EXP_MANTISSA_BMSK) == 0);

    if (is_zero) {
        return 0;
    }

    uint32_t mantissa = input & 0x007fffff;
    uint32_t exp = (input & 0x7f800000) >> 23;
    uint32_t sign = (input & 0x80000000) >> 31;

    if (is_exp_a) {
        int32_t se = static_cast<int32_t>(exp);
        // rebias
        se = se - 127 + 15;
        // check for saturation
        if (se > 31) {
            se = 31;
            mantissa = 0x007fffff;
        } else if (se < 0) {
            se = 0;
            mantissa = 0x0;
        }

        exp = static_cast<uint32_t>(se);
    }

    // float mantissa is 23 bits + hidden bit = 24 bits
    // add hidden 1
    mantissa = (1 << 23) | mantissa;

    if (shared_exp >= exp) {
        int exp_diff = shared_exp - exp;
        // shift mantissa further down by exp diff
        // In bit-shift operation (A >> B), the result is undefined if B is greater than or equal to the number of bits
        // in A
        while (exp_diff > 31) {
            mantissa = mantissa >> 31;
            exp_diff -= 31;
        }
        mantissa = mantissa >> exp_diff;
    }

    // this needs to become 7 bits so shift 17 times
    if (truncate_bfp_mantissa) {
        // Truncation: Round down
        mantissa = mantissa >> 17;
    } else {
        // Round mantissa to nearest even
        mantissa += 1 << 16;       
        mantissa = mantissa >> 17;
        if (mantissa > 127) {
            mantissa = 127;
        }
    }

    // add sign bit only if result is not 0
    if (0 == mantissa) {
        sign = 0;
    }
    mantissa = (sign << 7) | mantissa;
    return mantissa;
}

inline uint32_t convert_bfp_to_u32(uint8_t data, uint8_t shared_exp, bool is_exp_a) {
    uint32_t exp = shared_exp;
    uint32_t out_num = 0;

    uint32_t sign = data >> 7;
    uint32_t man = data & 0x7f;

    // Shift mantissa up until there is a 1 in bit 6
    int shift_cnt = 0;
    if (man == 0) {
        man = 0;
        exp = 0;
    } else {
        // shift_cnt = 6 - (31 - __builtin_clz(man));
        // man = (man << (shift_cnt + 1)) & 0x7f;
        while ((man & 0x40) == 0) {
            man = man << 1;
            shift_cnt++;
        }
        // shift one more time and zero the
        // hidden top mantissa bit
        // shift again to put first non-hidden mantissa
        // bit in bit 7
        man = man << 1;
        man = man & 0x7f;

        // adjust exponent
        exp = exp - shift_cnt;

        // if exp_a rebias exp to 127
        if (is_exp_a) {
            exp = exp - 15 + 127;
        }
    }
    std::cout << std::hex << "mantissa without hidden bit: " << man << std::endl;

    // put s, e, m together
    out_num = (sign << 31) | (exp << 23) | (man << 16);
    
    return out_num;
}

int main() {
    float input = 64.5;
    uint32_t uint32_num = *reinterpret_cast<uint32_t*>(&input);
    uint32_t shared_exp = uint32_num >> 23;
    std::cout << std::hex << "shared exp: " << shared_exp << std::endl;

    uint32_t mantissa = convert_u32_to_bfp8(uint32_num, shared_exp, false);
    std::cout << std::hex << "mantissa with hidden bit: " << mantissa << std::endl;

    uint32_t out_uint32 = convert_bfp_to_u32(mantissa, shared_exp, false);
    float out = *reinterpret_cast<float*>(&out_uint32);
    std::cout << "round trip output: " << out << std::endl;
}
```


To me there might be a bug in the line `man = man << 1` from `convert_bfp_to_u32`:
```
        man = man << 1;
        man = man & 0x7f;
```
If we subtract 1 here, we get back the expected output, but I am not familiar enough with this code to draw any conclusions. @Davor @abhullar-tt Thoughts?

**Comment by TT-BrianLiu** [comment]
Btw I use `uint32_t` instead of proper `uint8_t` for the returned exponents and mantissa because godbolt wasn't printing correctly for me, but it shouldn't affect the code.

**Comment by razorback3** [comment]
>If we subtract 1 here, we get back the expected output, but I am not familiar enough with this code to draw any conclusions. @davor @abhullar-tt Thoughts?

So, the conclusion is that existing ttnn code has bug?

**Comment by zzigler-tt** [comment]
@abhullar-tt Can you please review and provide feedback? TY

**Comment by hschoi4448** [comment]
>@abhullar-tt Can you please review and provide feedback? TY

I don't see any related PR or commit. Is there any progress on this?
@zzigler-tt

**Comment by abhullar-tt** [comment]
@TT-BrianLiu for the first example, it seems like the expected answer can be achieved if `convert_u32_to_bfp8` with `truncate_bfp_mantissa=true`

I still need to look into the second case

**Comment by TT-BrianLiu** [comment]
> [@TT-BrianLiu](https://github.com/TT-BrianLiu) for the first example, it seems like the expected answer can be achieved if `convert_u32_to_bfp8` with `truncate_bfp_mantissa=true`
> 
> I still need to look into the second case

That's true. Maybe `convert_bfp_to_u32` is default always using `truncate_bfp_mantissa=true` and needs to respect the other mode?

**Comment by zzigler-tt** [comment]
Hello @abhullar-tt @TT-BrianLiu checking in on the progress of this issue as it's a high priority item. Please advise on projected resolution timeframe so we can level set with impacted folks. Thanks

**Comment by TT-BrianLiu** [comment]
I can take a look at this next week. Likely it involves adding a `truncate_bfp_mantissa=false` mode to `convert_bfp_to_u32` since this function is basically default assuming `truncate_bfp_mantissa=true`. But flipping to round instead of truncating in `convert_bfp_to_u32` might have consequences to PCC across the stack, so it's hard to gauge the time it would take to fix this.

**Comment by hschoi4448** [comment]
> Regarding your second test case with incorrect handling of overflow, that's expected. With shared exponents, if we have overflow we need to recalculate the shared exponent and renormalize all the datums that use the shared exponents. So it's not so easy to fix.

Upon hearing that an overflow would require recalculating 16 values, I wondered if it was really necessary to handle this process. However, upon reviewing the existing implementation, I found that `get_max_exp` is already iterating through 16 elements to find `max_exp`.

It seems that overflow would only occur in `max(abs(16 elements))`. Wouldn’t it be sufficient to just add an additional overflow check there? What do you think?

@TT-BrianLiu @razorback3

**Comment by TT-BrianLiu** [comment]
I merged PR (https://github.com/tenstorrent/tt-metal/pull/16997) to properly round mantissa.

> It seems that overflow would only occur in max(abs(16 elements)). Wouldn’t it be sufficient to just add an additional overflow check there? What do you think?

Yea, it's possible to implement, but it would require slightly more involved modification of the conversion process. Ideally, process and pack 16 elements at the same time to avoid having to recalculate rounded mantissas. @hschoi4448 Is this a high priority item for Moreh?

**Comment by rdjogoTT** [comment]
@TT-BrianLiu @hschoi4448 

We just confirmed with the HW team that WH and BH both DO NOT support rounding to even, only rounding up in case of ties. This means the original implementation was accurately modelling HW.

Regarding the mantissa overflow case, this is also correctly modelling what the HW does in this case. The HW also would not increment the exponent, but round down to 7.9375 instead.

In both cases we are dealing with HW limitations for deterministic rounding. If you want more even rounding, you would have to sacrifice determinism and use stochastic rounding.

**Comment by hschoi4448** [comment]
>We just confirmed with the HW team that WH and BH both DO NOT support rounding to even, only rounding up in case of ties.

Looking at the implementation of `ttnn`'s unit tests, I assumed that `ttnn` aims to replicate the same behavior as PyTorch.

If the goal of ttnn is not to replicate PyTorch's behavior, could you clarify what its purpose is?

@rdjogoTT @TT-BrianLiu @razorback3

**Comment by TT-BrianLiu** [comment]
Goal of ttnn is not to match hardware. Rounding to nearest even on ties is better and should be done like this on host. For example, we still use host conversion to tilize and cache bfp8 weights, so the PR definitely has value. 

@rdjogoTT You also mentioned that Quasar is looking to fix this to match spec?

**Comment by rdjogoTT** [comment]
Yes, Quasar will have rounding ties to even, as well as stochastic rounding.

**Comment by razorback3** [comment]
Folks, thanks for the changes and comments!
Closing the issue as completed :)
